src/liana/method/sc/_compute_global_score.py

```diff
@@ -0,0 +1,217 @@
+import numpy as np
+import pandas as pd
+from scipy.sparse import issparse, csr_matrix, diags
+from anndata import AnnData
+from joblib import Parallel, delayed
+
+# --- Placeholder for li.mt._constants.DefaultValues ---
+class DefaultValues:
+    lr_sep = '^'
+V = DefaultValues()
+
+# =================================================================
+# 1. Helper Function: Splits complex names
+# =================================================================
+
+def _split_complex(name: str, complex_sep: str = "_"):
+    """Helper for splitting complex names."""
+    toks = name.split(complex_sep)
+    if len(toks) > 1:
+        return toks[0], complex_sep.join(toks[1:])
+    else:
+        return name, name
+
+# =================================================================
+# 2. Worker Function: Computes score for a single permutation
+# =================================================================
+
+def _run_single_permutation(
+    shuffled_labels: np.ndarray,
+    lrdata_X,
+    interaction_names: np.ndarray,
+    celltype_names: list,
+    xy_sep: str,
+) -> dict:
+    """Run a single permutation using precomputed shuffled labels."""
+    cats = pd.Categorical(shuffled_labels, categories=celltype_names, ordered=False)
+    col_idx = cats.codes
+
+    n_cells = col_idx.size
+    n_types = len(celltype_names)
+    row_idx = np.arange(n_cells)
+
+    # Build sparse one-hot and normalize columns
+    t_sparse = csr_matrix(
+        (np.ones(n_cells, dtype=np.float64), (row_idx, col_idx)),
+        shape=(n_cells, n_types)
+    )
+    col_sums = np.asarray(t_sparse.sum(axis=0)).ravel()
+    col_sums[col_sums == 0] = 1.0
+    t_sparse = t_sparse @ diags(1.0 / col_sums)
+
+    # Ensure X is sparse CSR
+    X = lrdata_X if issparse(lrdata_X) else csr_matrix(lrdata_X)
+
+    # Aggregation
+    result = t_sparse.T @ X
+    values = (result.toarray() if issparse(result) else np.asarray(result)).ravel()
+
+    # Build keys with vectorized string ops
+    interaction_names_tiled = np.tile(interaction_names.astype(str), n_types)
+    celltype_names_repeated = np.repeat(
+        np.asarray(celltype_names, dtype=str),
+        interaction_names.shape[0]
+    )
+    keys = np.char.add(np.char.add(interaction_names_tiled, xy_sep), celltype_names_repeated)
+
+    return dict(zip(keys.tolist(), values.tolist()))
+
+# =================================================================
+# 3. Main Function: Computes score and p-value
+# =================================================================
+
+def compute_global_score(
+    lrdata: AnnData,
```

Comment:
This function is applicable to any adata object, not just lrdata - lets keep it more general